### PR TITLE
feat: bootstrap cloud functions

### PR DESCRIPTION
**Pull request by xhanin**
get work from my initial poc

slight review to integrate with shared data structures, including very basic new properties fetching

### CODE REVIEW

File: cloud/functions/package.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "functions",
+  "scripts": {
+    "lint": "eslint --ext .js,.ts .",
+    "build": "tsc",
+    "build:watch": "tsc --watch",
+    "serve": "npm run build && firebase emulators:start  --import data --export-on-exit",
+    "shell": "npm run build && firebase functions:shell",
```
Comment (fcamblor) [inline cloud/functions/package.json:8]: @xhanin when I run `npm run shell` I get a `firebase: command not found`
=> I assume there is a missing `firebase` dev dependency (you may have installed `firebase` globally ?)

Comment (xhanin) [reply cloud/functions/package.json:8]: I've never used it actually, it was generated by the initial command. I remove it.

Comment (fcamblor) [reply cloud/functions/package.json:8]: fixed through 756ccf7 (I rolled back your changes on b3d56e0 as I made a mistake : it's not `firebase` but `firebase-tools` which was missing0

File: cloud/functions/src/crawlers/crawl.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    const snapshot = await db.collection("crawlers/devoxx/events").where("crawl", "==", true).get();
+    if (snapshot.empty) {
+        info("no events to crawl")
+    } else {
+        for (const doc of snapshot.docs) {
+            info("crawling devoxx event " + doc.id)
+            const event = await crawlDevoxx(doc.id)
+            await saveEvent(event)
+            events.push({id: event.id})
+        }
```
Comment (fcamblor) [inline cloud/functions/src/crawlers/crawl.ts:21]:
```suggestion
        await Promise.all(snapshot.docs.map(async doc => {
            info("crawling devoxx event " + doc.id)
            const event = await crawlDevoxx(doc.id)
            await saveEvent(event)
            events.push({id: event.id})
        }))
```
Let's parallelize crawlings (went from 29s to 19s on my laptop)

File: cloud/functions/src/crawlers/crawl.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    for (const talksStat of event.talkStats) {
+        // TODO: see if we really want to override stats each time we crawl
+        info("saving stats " + talksStat);
+        for (const talkStat of talksStat.stats) {
+            await db.collection("events").doc(event.id)
+            .collection("days").doc(talksStat.day)
+            .collection("talksStats").doc(talkStat.id)
+            .set(talkStat)
+        }
+    }
```
Comment (fcamblor) [inline cloud/functions/src/crawlers/crawl.ts:55]:
```suggestion
    const firestoreEvent = await db.collection("events").doc(event.id);

    await Promise.all(event.daySchedules.map(async daySchedule => {
        await firestoreEvent
            .collection("days").doc(daySchedule.day)
            .set(daySchedule)
    }))

    await Promise.all(event.talks.map(async talk => {
        info("saving talk " + talk.id + " " + talk.title);
        await firestoreEvent
            .collection("talks").doc(talk.id)
            .set(talk)
    }));

    await Promise.all(
        event.talkStats
            .flatMap(dailyTalksStat =>
                dailyTalksStat.stats.map(stat => ({day: dailyTalksStat.day, stat: stat }))
            ).map(async talkStatWithDay => {
                // TODO: see if we really want to override stats each time we crawl
                await firestoreEvent
                    .collection("days").doc(talkStatWithDay.day)
                    .collection("talksStats").doc(talkStatWithDay.stat.id)
                    .set(talkStatWithDay.stat)
            })
    )
```
Parallelize everything, went from 19s => 1s :-)

File: cloud/functions/src/crawlers/devoxx/crawler.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+    const event: FullEvent = { id: eventId, info: eventInfo, daySchedules: [], talkStats: [], talks: []}
+    for (const day of days) {
+        const {daySchedule, talkStats, talks} = await crawlDevoxxDay(eventId, day.id)
+        event.daySchedules.push(daySchedule)        
+        event.talkStats.push({day: day.id, stats: talkStats})
+        for (const talk of talks) {
+            event.talks.push(talk)
+        }
+    }
```
Comment (fcamblor) [inline cloud/functions/src/crawlers/devoxx/crawler.ts:66]:
```suggestion
    await Promise.all(days.map(async day => {
        const {daySchedule, talkStats, talks} = await crawlDevoxxDay(eventId, day.id)
        event.daySchedules.push(daySchedule)
        event.talkStats.push({day: day.id, stats: talkStats})
        for (const talk of talks) {
            event.talks.push(talk)
        }
    }))
```
Let's parallelize days : went from 19s to 17s by doing this locally

**Comment by fcamblor** [comment]
Merging to move things forward (I'll parallelize stuff in a separate PR)

**Pull request merged**